File: src/main/java/com/nanaios/polygonal_tech/capability/interfaces/ILongFluidHandler.java
package com.nanaios.polygonal_tech.capability.interfaces;

import com.nanaios.polygonal_tech.fluids.base.LongFluidStack;
import com.nanaios.polygonal_tech.util.MathUtil;
import net.minecraftforge.fluids.FluidStack;
import net.minecraftforge.fluids.capability.IFluidHandler;
import org.jetbrains.annotations.NotNull;

/// long型の流体量を扱うためのインターフェース
public interface ILongFluidHandler extends IFluidHandler {
    /// タンク内の流体をLongFluidStackで取得します。
    ///
    /// @param tank タンクのインデックス
    /// @return タンク内の流体をLongFluidStackで返します。
    @Override
    @NotNull LongFluidStack getFluidInTank(int tank);

    /// タンクの最大容量をlong型で取得します。
    ///
    /// @param tank タンクのインデックス
    /// @return タンクの最大容量をlong型で返します。
    long getTankLongCapacity(int tank);

    /// 流体を抽出します。
    ///
    /// @param resource 抽出する流体の種類と量をLongFluidStackで指定します。
    /// @param action   抽出のシミュレーションを行うかどうかを指定します。
    /// @return 実際に抽出された流体の種類と量をLongFluidStackで返します。
    @NotNull LongFluidStack drainLong(LongFluidStack resource, FluidAction action);

    /// 流体を抽出します。
    ///
    /// @param maxDrain 抽出する流体の最大量をlong型で指定します。
    /// @param action   抽出のシミュレーションを行うかどうかを指定します。
    /// @return 実際に抽出された流体の種類と量をLongFluidStackで返します。
    @NotNull LongFluidStack drainLong(long maxDrain, FluidAction action);

    /// 流体を受け取ります。
    ///
    /// @param resource 受け取る流体の種類と量をLongFluidStackで指定します。
    /// @param action   受け取りのシミュレーションを行うかどうかを指定します。
    /// @return 実際に受け取った流体の量をlong型で返します。
    long fillLong(LongFluidStack resource, FluidAction action);


    @Override
    default int fill(FluidStack resource, FluidAction action) {
        return MathUtil.longToInt(fillLong(LongFluidStack.from(resource), action));
    }

    @Override
    @NotNull
    default LongFluidStack drain(FluidStack resource, FluidAction action) {
        return drainLong(LongFluidStack.from(resource), action);
    }

    @Override
    @NotNull
    default LongFluidStack drain(int maxDrain, FluidAction action) {
        return drainLong(maxDrain, action);
    }

    @Override
    default int getTankCapacity(int tank) {
        return MathUtil.longToInt(getTankLongCapacity(tank));
    }
}

File: src/main/kotlin/com/nanaios/polygonal_tech/core/capability/fluid/ILongFluidHandler.kt
package com.nanaios.polygonal_tech.core.capability.fluid

import com.nanaios.polygonal_tech.core.fluid.LongFluidStack
import com.nanaios.polygonal_tech.core.fluid.toLongFluidStack
import com.nanaios.polygonal_tech.core.util.roundInt
import net.minecraft.commands.arguments.ResourceArgument.resource
import net.minecraftforge.fluids.FluidStack
import net.minecraftforge.fluids.capability.IFluidHandler

/**
 * 複数の[ILongFluidTank]を統括して操作し、外部（パイプ等）との大容量流体（[LongFluidStack]）の授受を
 * まとめて処理できるようにすることを目的としたCapabilityインターフェース。
 */
interface ILongFluidHandler: IFluidHandler {
    /** このハンドラーが管理している全タンクのリスト */
    val tanks: List<ILongFluidTank>

    /**
     * 所持している複数のタンクに対して順次流体の注入を試み、その合計注入量を返すことを目的とするメソッド。
     *
     * @param resource 注入したい流体
     * @param action 実際の更新かシミュレートかを示す[IFluidHandler.FluidAction]
     * @return 実際に注入された総量
     */
    fun fill(resource: LongFluidStack, action: IFluidHandler.FluidAction): Long {
        var total = 0L
        tanks.forEach { tank ->
            val filled = tank.fill(resource,action)
            total += filled
            resource.shrink(filled)
        }
        return total
    }

    /**
     * 該当する流体の種類と量を指定して、保持しているタンク群から引き出すことを目的とするメソッド。
     * 内部的には要求量だけを指定する[drain]に委譲される。
     *
     * @param resource 引き出したい流体の種類と量を指定する[LongFluidStack]
     * @param action 実際の更新かシミュレートかを示す[IFluidHandler.FluidAction]
     * @return 引き出された流体と量のまとまり（[LongFluidStack]）
     */
    fun drain(resource: LongFluidStack, action: IFluidHandler.FluidAction): LongFluidStack = drain(resource.getAmount(), action)

    /**
     * 所持しているタンク群から順次探索し、指定された最大量まで同種の流体を引き出してまとめることを目的とするメソッド。
     *
     * @param maxDrain 引き出したい最大[Long]量
     * @param action 実際の更新かシミュレートかを示す[IFluidHandler.FluidAction]
     * @return 引き出された流体と量のまとまり（[LongFluidStack]）
     */
    fun drain(maxDrain: Long, action: IFluidHandler.FluidAction): LongFluidStack {
        var drained:LongFluidStack = LongFluidStack.EMPTY
        var want = maxDrain

        tanks.forEach { tank ->
            if(drained.isEmpty()){
                // 最初に引き出せるタンクを見つけて、流体を引き出す。
                drained = tank.drain(want, action)
                want -= drained.getAmount()
            } else if(tank.longFluid.isFluidEqual(drained)) {
                // すでに引き出した流体と同じ流体が入っているタンクを見つけたら、さらに引き出す。
                val additional = tank.drain(want, action).getAmount()
                drained.grow(additional)
                want -= additional
            }
        }

        return drained
    }

    fun isFluidValid(tank: Int, stack: LongFluidStack): Boolean = tanks[tank].isFluidValid(stack)
    fun getLongFluidInTank(tank: Int): LongFluidStack = tanks[tank].longFluid
    fun getTankLongCapacity(tank: Int): Long = tanks[tank].longCapability

    override fun getTanks(): Int = tanks.size
    override fun getFluidInTank(tank: Int) = getLongFluidInTank(tank).toFluidStack()
    override fun getTankCapacity(tank: Int) = getTankLongCapacity(tank).roundInt()
    override fun isFluidValid(tank: Int, stack: FluidStack) = isFluidValid(tank, stack.toLongFluidStack())
    override fun fill(resource: FluidStack, action: IFluidHandler.FluidAction) = fill(resource.toLongFluidStack(), action).roundInt()
    override fun drain(maxDrain: Int, action: IFluidHandler.FluidAction) = drain(maxDrain.toLong(), action).toFluidStack()
    override fun drain(resource: FluidStack, action: IFluidHandler.FluidAction) = drain(resource.toLongFluidStack(), action).toFluidStack()
}